### PR TITLE
test(e2e): fail loudly when the harness runs Obsidian below minAppVersion

### PR DESCRIPTION
**Pull request by chhoumann**
## TL;DR

This branch began as a compatibility audit: a worker e2e-validating PR #1438 reported that `ButtonComponent.setDestructive` "appears ABSENT on Obsidian 1.13.0", crashing `AIAssistantProvidersModal` and `GenericYesNoPrompt` on QuickAdd's declared `minAppVersion`.

**The premise is refuted.** `setDestructive` is `@since 1.13.0` and is present in the real Obsidian 1.13.0 build. **No QuickAdd plugin code is changed.** The real defect is in the **e2e harness**: it could silently run a *sub-minimum* Obsidian build and mislabel it, which is exactly what produced the false "missing API" signal. This PR fixes that with a pre-launch `minAppVersion` guard.

## Part 1 - The audit: no real compatibility gap

Obsidian ships two versions: the **installer shell** (the `.app`, shown in the window title) and the auto-updated **app code** (`obsidian-<version>.asar`, == the `apiVersion` plugins see). `minAppVersion` is enforced against the **app-code** version. On the worker's machine the installer was `1.12.7`; with no `>= 1.13.0` asar reachable, Obsidian fell back to the bundled `1.12.7` build - which genuinely lacks `setDestructive`. That sub-minimum app, not a real 1.13.0, is what crashed.

Four independent sources agree `setDestructive` lands at **1.13.0**:

| Source | `setDestructive`? |
|---|---|
| Real **1.13.0** asar (`package.json: "1.13.0"`) | ✅ present (`prototype.setDestructive=`, 15×) |
| **1.12.7** asar (installer fallback) | ❌ absent (only deprecated `setWarning`) |
| Typings `obsidian.d.ts` | ✅ `@since 1.13.0` (`setWarning` now `@deprecated`) |
| Live isolated **1.13.0** e2e instance | ✅ `api.yesNoPrompt(...)` constructed, no throw |

Full sweep (`src/` incl. `src/gui`): the newest Obsidian API QuickAdd uses is `@since 1.13.0` (`SettingDefinitionGroup`/`SettingDefinitionItem` and `setDestructive`) - which is *why* `minAppVersion` is 1.13.0. The only API members `@since > 1.13.0` in the typings are all `@since 1.13.1` declarative-settings additions (`DisplayValueComponent`, `addDisplayValue`, `setStatus`, `SettingDefinition.displayValue/status/displayFormat/search/match/placeholder`) - **none used**. Every `@ts-ignore`/`as any`/`.svelte` runtime path reaches only ancient untyped APIs (`metadataCache.getTags`, `app.commands.commands`, `String.contains`). An adversarial refutation panel (4 lenses, every claim grep-verified against the real 1.13.0 binary) found **0 gaps**.

**Conclusion: `minAppVersion: 1.13.0` is accurate and is the lowest honest value. No plugin code change is warranted.**

## Part 2 - The fix: the harness must not run below minAppVersion

The e2e harness launches the locally-installed Obsidian via `open -a`. Obsidian loads the newest auto-updated `obsidian-*.asar` from the **real login-user config dir** (resolved via `getpwuid`, so the harness's isolated `$HOME`/`--user-data-dir` do not affect it), floored at the bundled installer asar. When nothing reaches `minAppVersion`, it silently runs the bundled installer build - which can be **below** `minAppVersion`. e2e then runs against an app QuickAdd does not support, turning every genuine "missing API" result into a **false** crash signal. That is the trap that started this whole investigation.

`assertObsidianMeetsMinAppVersion` (in `scripts/start-obsidian-e2e-instance.mjs`, wired into both entry points: `start` `main()` and the `obsidian:e2e` CLI) now, **before launch**:
- resolves the app-code version the way Obsidian does - newest **valid** installed asar (validated by reading its own `package.json`, so a partial/corrupt download is skipped rather than trusted by filename), floored at the bundled installer asar;
- aborts with a clear, actionable error when it is below the worktree manifest's `minAppVersion`, before spawning a doomed instance;
- surfaces the resolved version in the ready banner / CLI stderr.

### Live red/green

```
# GREEN (this machine: real app-code 1.13.0)
$ pnpm run obsidian:e2e -- eval code='"green"'
Obsidian app 1.13.0 (installer 1.12.7), plugin minAppVersion 1.13.0
=> green

# RED (sub-minimum build, the worker's scenario)
Obsidian app version 1.12.7 is BELOW the plugin's minAppVersion 1.13.0 (installer
shell 1.12.7). Obsidian fell back to a build QuickAdd does not support, so any e2e
"missing API" failure here is a FALSE signal, not a real bug. Update Obsidian to
>= 1.13.0 ... before running e2e.
```

## Design notes / why predict from disk

Obsidian does not expose `apiVersion` to a CDP-evaluated snippet (`require("obsidian")` is unresolvable outside plugin scope; the window title carries the *installer* version), so the app-code version cannot be read from the running renderer. The guard therefore resolves it the same way Obsidian does and runs **before** launch, so a fresh instance loads exactly what was resolved. Documented residual: a warm instance reused from an earlier run reflects the asar resolution at *its* launch (an Obsidian update mid-session is not re-detected) - bounded, since instances are reaped per worktree and Obsidian versions are stable within a run.

This was adversarially reviewed (opposing model); the HIGH findings (post-launch placement, filename-trust of cached asars) are addressed here (pre-launch guard, asar content validation + a corrupt-asar test).

## Gates
`tsc -noEmit` ✅ · `eslint .` ✅ · `svelte-check` ✅ 0 errors · `vitest` ✅ 3326 passed / 37 skipped (+8 new guard tests: green / red / corrupt-asar skip / undeterminable refusal).

No conflict with #1438 - this PR touches only `scripts/` + `tests/`, not `AIAssistantProvidersModal.ts`.

## Summary by CodeRabbit

* **New Features**
  * Added a minimum Obsidian app-code version check before starting or reusing an end-to-end instance.
  * Enhanced non-JSON output to show the resolved Obsidian app version together with the plugin’s required `minAppVersion` when available.
* **Bug Fixes**
  * Improved app-code version resolution by inspecting installed archives, selecting the best candidate, and handling malformed/corrupt archives with clear failures.
  * Strengthened version comparison to correctly order major/minor/patch values and enforce minimum requirements.
* **Tests**
  * Added coverage for version comparison, archive-based app-version resolution (including edge-case alignment), and min-version enforcement/rejection behavior.

### CODE REVIEW

File: scripts/start-obsidian-e2e-instance.mjs [hunk 1]
```diff
@@ -393,6 +394,181 @@ export async function trustVaultAndVerifyQuickAdd(options) {
...
+		const headerBuf = Buffer.alloc(headerSize);
+		await handle.read(headerBuf, 0, headerSize, 16);
+		const header = JSON.parse(headerBuf.toString("utf8"));
+		const entry = header?.files?.["package.json"];
+		if (!entry) return null;
+		const size = Number(entry.size);
+		const offset = Number(entry.offset);
+		if (!Number.isFinite(size) || !Number.isFinite(offset) || size <= 0) return null;
+		const fileBuf = Buffer.alloc(size);
+		await handle.read(fileBuf, 0, size, 16 + headerSize + offset);
```
[the remark this answers is not in the document]
Comment (chhoumann) [reply scripts/start-obsidian-e2e-instance.mjs]: Confirmed and fixed in 94caeede. You're right: the header is a Pickle-wrapped, 4-byte-aligned JSON string, so file offsets are relative to `8 + headerPickleLength` (byte 4), not `16 + jsonLength`. I reproduced it with a spec-correct asar whose header JSON length is not 4-aligned (`pad=1`): the old reader returned `null`, ignoring a valid `obsidian-*.asar`. It only worked here because the real 1.13.0/1.12.7 asar headers happened to be 4-aligned.

Fix uses the header pickle length for the data base. The test `buildAsar` is now a spec-correct Pickle layout, and a new test reads the version at all four 4-byte residues (the old math fails 3 of 4). Live guard still resolves the real `obsidian-1.13.0.asar` correctly.

**Comment by chhoumann** [comment]
Addressed the Codex P2 (asar header offset) in 94caeede - details in the inline thread.

Re: CodeRabbit's nitpick to route the compatibility note through a `logger` instead of `console.error` (`obsidian-e2e-cli.mjs:134`): the `logger` guideline targets the plugin source under `src/` (that logger surfaces via `Notice` in the Obsidian runtime). The `scripts/*.mjs` e2e harness CLIs have no logger and use `console.*` exclusively - e.g. `start-obsidian-e2e-instance.mjs` and `stop-obsidian-e2e-instance.mjs` already log via `console.error`/`console.log`, and the reaper is passed `log: console.error`. Using `console.error` here (stderr, keeping command stdout clean) matches the file's established convention, so I've kept it.

**Comment by chhoumann** [comment]
@codex review

File: tests/start-obsidian-e2e-instance.test.ts [hunk 2]
```diff
@@ -197,6 +200,168 @@ describe("assertSecureDirIfPresent", () => {
...
+// pickle][file data], with the header JSON 4-byte aligned inside the header
+// pickle). `extraJsonPad` widens the header JSON so its length lands on a chosen
+// 4-byte residue, exercising the alignment padding that a naive `16 + jsonLength`
+// reader gets wrong.
+function buildAsar(version: string, extraJsonPad = 0): Buffer {
+	const pkg = Buffer.from(JSON.stringify({ version }), "utf8");
+	const files: Record<string, unknown> = {
+		"package.json": { offset: "0", size: pkg.length },
+	};
+	if (extraJsonPad > 0) files._pad = "x".repeat(extraJsonPad);
```
[the remark this answers is not in the document]
Comment (chhoumann) [reply tests/start-obsidian-e2e-instance.test.ts]: Good catch - fixed in beff53db. `_pad` is now always present, so each `extraJsonPad` step shifts the header JSON length by exactly one byte; the four builds span residues {0,1,2,3}, and the test now asserts that set explicitly so coverage can't silently collapse again. Verified: old buildAsar covered residues [3,2,3,0] (3 distinct), now [1,2,3,0] (4 distinct).
